Allow sharing UTS namespace of another container

Rid opened the pull request:
Signed-off-by: Grant Millar <rid@cylo.io>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Allow sharing of the UTS namespace of another container using `--uts container:<name|id>`

**- How I did it**
- Updated `api/types/container/host_config.go` to allow for checking the parent container
- Added the getUtsContainer function to `daemon/container_operations_unix.go`
- Added the namespace logic to `daemon/oci_linux.go`

**- How to verify it**
```
# docker run -d --rm \
--name foo_sandbox \
--cgroup-parent pod_foo.slice \
--ipc 'shareable' \
alpine sleep infinity

# docker run -d --rm \
--name app \
--cgroup-parent pod_foo.slice \
--network container:foo_sandbox \
--ipc container:foo_sandbox \
--uts container:foo_sandbox \
kennethreitz/httpbin

# lsns | grep -E 'sleep|httpbin'
4026533142 mnt         1 963921 root             sleep infinity
4026533143 uts         3 963921 root             sleep infinity
4026533144 ipc         3 963921 root             sleep infinity
4026533145 pid         1 963921 root             sleep infinity
4026533147 net         3 963921 root             sleep infinity
4026533203 cgroup      1 963921 root             sleep infinity
4026533205 mnt         2 964606 root             /usr/bin/python3 /usr/local/bin/gunicorn -b 0.0.0.0:80 httpbin:app -k gevent
4026533206 pid         2 964606 root             /usr/bin/python3 /usr/local/bin/gunicorn -b 0.0.0.0:80 httpbin:app -k gevent
4026533207 cgroup      2 964606 root             /usr/bin/python3 /usr/local/bin/gunicorn -b 0.0.0.0:80 httpbin:app -k gevent

# ls -l /proc/963921/ns
total 0
lrwxrwxrwx 1 root root 0 Mar  8 12:59 cgroup -> 'cgroup:[4026533203]'
lrwxrwxrwx 1 root root 0 Mar  8 12:58 ipc -> 'ipc:[4026533144]'
lrwxrwxrwx 1 root root 0 Mar  8 12:59 mnt -> 'mnt:[4026533142]'
lrwxrwxrwx 1 root root 0 Mar  8 12:58 net -> 'net:[4026533147]'
lrwxrwxrwx 1 root root 0 Mar  8 12:59 pid -> 'pid:[4026533145]'
lrwxrwxrwx 1 root root 0 Mar  8 13:00 pid_for_children -> 'pid:[4026533145]'
lrwxrwxrwx 1 root root 0 Mar  8 12:59 user -> 'user:[4026531837]'
lrwxrwxrwx 1 root root 0 Mar  8 12:58 uts -> 'uts:[4026533143]'

# ls -l /proc/964606/ns
total 0
lrwxrwxrwx 1 root root 0 Mar  8 12:59 cgroup -> 'cgroup:[4026533207]'
lrwxrwxrwx 1 root root 0 Mar  8 12:59 ipc -> 'ipc:[4026533144]'
lrwxrwxrwx 1 root root 0 Mar  8 12:59 mnt -> 'mnt:[4026533205]'
lrwxrwxrwx 1 root root 0 Mar  8 12:59 net -> 'net:[4026533147]'
lrwxrwxrwx 1 root root 0 Mar  8 12:59 pid -> 'pid:[4026533206]'
lrwxrwxrwx 1 root root 0 Mar  8 13:01 pid_for_children -> 'pid:[4026533206]'
lrwxrwxrwx 1 root root 0 Mar  8 12:59 user -> 'user:[4026531837]'
lrwxrwxrwx 1 root root 0 Mar  8 12:59 uts -> 'uts:[4026533143]'

# docker exec foo_sandbox hostname
223bae70638b

# docker exec app hostname
223bae70638b
```


**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Allow sharing UTS namespace of another container

**- A picture of a cute animal (not mandatory but encouraged)**
🐤

This will also require changes to the CLI to support the new functionality which I will submit a PR for shortly, the PR for the CLI will require vendoring from the API as it requires changes to `host_config` interfaces.

tianon commented:
I'm really curious about the underlying use case for this -- isn't the UTS namespace _just_ the hostname and domain name?  What does sharing those with another container get you that you don't get by just using the same values?

Along a similar vein, wouldn't doing `--uts container:xxx` have to also imply that setting either of `--hostname` and `--domainname` becomes an error, since we don't want the new container's creation to modify the state of the UTS namespace of the original container inadvertently?  (As in, this would also need to suppress any default values/behavior.)

Rid commented:
@tianon We currently run thousands of single containers with multiple services using s6 as the services are required to be inside the same container. When using the same ipc, net & uts namespaces we can leverage multiple containers acting as one, it will allow us to update parts of the container group without stopping the whole.

This is actually how Kubernetes is operating, each pod shares ipc, net & uts namespaces. Of course you could probably set the same hostname/domainname on each, but sharing the namespace is cleaner.

I have updated the commit to include errors when specifying `--uts container:xxx` with either `--hostname` or `--domainname`.

I've also added tests for each scenario.

tianon commented on daemon/oci_linux.go:
```diff
@@ -314,9 +314,30 @@ func WithNamespaces(daemon *Daemon, c *container.Container) coci.SpecOpts {
+				Path: fmt.Sprintf("/proc/%d/ns/uts", uc.State.GetPID()),
+			}
+			setNamespace(s, ns)
+			if userNS {
+				// to share a UTS namespace, they must also share a user namespace
+				nsUser := specs.LinuxNamespace{
+					Type: "user",
+					Path: fmt.Sprintf("/proc/%d/ns/user", uc.State.GetPID()),
+				}
+				setNamespace(s, nsUser)
+			}
```
I see this is true for `pid` above -- is it definitely also true for `uts`?
(Sorry for not knowing -- my searching has failed me and I want to just make sure :sweat_smile:)

Rid replied:
It's not specifically true that any of the namespaces must share a user namespace, so I assumed this was an implementation opinion specific to docker, as all the other namespaces have the user namespace shared (ipc, net & pid).

I can remove it if you believe it wasn't intended to be an opinion on the implementation.

tianon commented on runconfig/hostconfig_unix.go:
```diff
@@ -30,9 +30,12 @@ func validateNetMode(c *container.Config, hc *container.HostConfig) error {
+	if (hc.UTSMode.IsHost() || hc.UTSMode.IsContainer()) && c.Hostname != "" {
 		return ErrConflictUTSHostname
 	}
+	if hc.UTSMode.IsContainer() && c.Domainname != "" {
```
I know it didn't even include a check for this at all previously (:heart:!!!!), but this one should include `IsHost` as well, right?

Rid replied:
Yes, I did actually have the `IsHost` in there, but I removed it as there was no test for a conflict between `IsHost` and `Domainname`, so I thought perhaps it was a valid combination.

Do you want me to add back the check?

Rid replied:
I can also add some tests for `IsHost` with `Domainname`/`Hostname`.

Rid replied:
@tianon I have added the error and added two tests to check for this.

tianon reviewed (commented):
Very nice, thank you for clarifying!  This makes sense to me, and I appreciate that you've added some useful (and straightforward!) tests.

(Just a few minor questions on the implementation details.)

Rid committed:
Allow sharing UTS namespace of another container

Signed-off-by: Grant Millar <rid@cylo.io>

tianon reviewed (approved):
IMO this looks great, thanks! :+1:

I appreciate you handling all the edge cases (even ones we weren't handling correctly before!) and writing a bunch of tests too.

neersighted commented:
@Rid Would you mind rebasing this? Things have changed significantly since this PR was started, and it would be better if you were the one to resolve the conflicts.
